File: 2021-blind/Q2.cpp
#include <bits/stdc++.h>
using namespace std;

map<string, int> freq[11];
vector<string> solution(vector<string> orders, vector<int> course) {
    // A. 코스 메뉴 조합 추출
    for(auto order : orders){
        sort(order.begin(), order.end());
        for(int i = 1; i < (1 << order.size()); i++)
        {
            string menu;
            /*for(int j = 0; j < order.size(); j++){
                if(i & (1<<j)) course += menu[j];
            }*/
            int tmp = i;
            for(int j = 0; j < order.size(); j++){
                if(tmp % 2) menu += order[j];
                tmp /= 2;
            }
            freq[menu.size()][menu]++;
        }
    }
    // B. 코스요리 메뉴 계산
    vector<string> ans;
    for(auto i : course){
        int mxfreq = 0;
        for(auto p : freq[i])
            mxfreq = max(mxfreq, p.second);
        if(mxfreq < 2) continue;
        for(auto p : freq[i]){
            if(p.second == mxfreq) ans.push_back(p.first);
        }
    }
    
    sort(ans.begin(),ans.end());
    return ans;
}

File: 2021-blind/Q2.py
def solution(orders, course):
    freq = [{} for i in range(11)]
    # A. 코스 메뉴 조합 추출
    for order in orders:
        order = ''.join(sorted(order))
        for i in range(1, 1 << len(order)):
            menu = ''
            #for j in range(len(order)):
            #    if i & (1 << j): menu += order[j]
            tmp = i
            for j in range(len(order)):
                if tmp % 2: menu += order[j]
                tmp //= 2
            if menu in freq[len(menu)]: freq[len(menu)][menu] += 1
            else: freq[len(menu)][menu] = 1
    # B. 코스요리 메뉴 계산
    ans = []
    for i in course:
        if not freq[i]: continue # 비어있을 경우 continue
        mxfreq = max(freq[i].values())
        if mxfreq < 2: continue
        for p in freq[i].items():
            if p[1] == mxfreq: ans.append(p[0])
    return sorted(ans)        
